File: examples/cpp/validate_merkle_path_sha2_256.cpp
#include <nil/crypto3/hash/algorithm/hash.hpp>
#include <nil/crypto3/hash/sha2.hpp>

using namespace nil::crypto3;

bool is_same(typename hashes::sha2<256>::block_type block0,
    typename hashes::sha2<256>::block_type block1){

    return block0[0] == block1[0] && block0[1] == block1[1];
}

[[circuit]] bool
    validate_path(std::array<typename hashes::sha2<256>::block_type, 0x05> merkle_path,
        typename hashes::sha2<256>::block_type leave,
        typename hashes::sha2<256>::block_type root) {
    
    typename hashes::sha2<256>::block_type subroot = leave;

    for (int i = 0; i < 0x05; i++) {
        subroot = hash<hashes::sha2<256>>(subroot, merkle_path[i]);
    }

    return is_same(subroot, root);
}

File: examples/rust/examples/validate_merkle_path_sha2_256.rs
#![no_main]

use std::intrinsics::{assigner_exit_check, assigner_sha2_256};

use ark_pallas::Fq;

use unroll::unroll_for_loops;

type Block = [Fq; 2];

#[circuit]
#[unroll_for_loops]
pub fn validate_path(merkle_path: [Block; 5], leave: Block, root: Block) {
    // TODO: (aleasims) use `fold` here instead of explicit iteration.
    // Right now assigner fails with `fold`.
    let mut subroot = leave;
    for i in 0..5 {
        subroot = hash(subroot, merkle_path[i]);
    }
    unsafe { assigner_exit_check(equal(subroot, root)) }
}

// TODO: (aleasims) remove this function and simply use `==` instead.
// Right now assigner fails if `==` is used.
fn equal(x: Block, y: Block) -> bool {
    x[0] == y[0] && x[1] == y[1]
}

// This will be SDK function in the future.
// All of the implementation details will be hidden from user.
fn hash(block1: Block, block2: Block) -> Block {
    let sha = assigner_sha2_256([block1[0].0, block1[1].0], [block2[0].0, block2[1].0]);
    [sha[0].into(), sha[1].into()]
}
